Sources/MongoSwift/BSON/BsonEncoderNew.swift

```diff
@@ -0,0 +1,647 @@
+        case dictionary(MutableDictionary, String)
+    }
+
+    /// The encoder we're referencing.
+    fileprivate let encoder: _BSONEncoder
+
+    /// The container reference itself.
+    private let reference: Reference
+
+    fileprivate init(referencing encoder: _BSONEncoder, at index: Int, wrapping array: MutableArray) {
+        self.encoder = encoder
+        self.reference = .array(array, index)
+        super.init(options: encoder.options, codingPath: encoder.codingPath)
+
+        self.codingPath.append(_BsonKey(index: index))
+    }
+
+    /// Initializes `self` by referencing the given dictionary container in the given encoder.
+    fileprivate init(referencing encoder: _BSONEncoder, key: CodingKey, wrapping dictionary: MutableDictionary) {
+        self.encoder = encoder
+        self.reference = .dictionary(dictionary, key.stringValue)
+        super.init(options: encoder.options, codingPath: encoder.codingPath)
+
+        self.codingPath.append(key)
+    }
+
+    fileprivate override var canEncodeNewValue: Bool {
+        // With a regular encoder, the storage and coding path grow together.
+        // A referencing encoder, however, inherits its parents coding path, as well as the key it was created for.
+        // We have to take this into account.
+        return self.storage.count == self.codingPath.count - self.encoder.codingPath.count - 1
+    }
+
+    /// Finalizes `self` by writing the contents of our storage to the referenced encoder's storage.
+    deinit {
+        let value: BsonValue?
+        switch self.storage.count {
+        case 0: value = nil
+        case 1: value = self.storage.popContainer()
+        default: fatalError("Referencing encoder deallocated with multiple containers on stack.")
+        }
+
+        switch self.reference {
+        case .array(let array, let index):
+            array.insert(value, at: index)
+
+        case .dictionary(let dictionary, let key):
+            dictionary[key] = value
+        }
+    }
+
+}
+
+/// Extend _BSONEncoder to add methods for "boxing" values.
+extension _BSONEncoder {
+
+    /// Converts a `CodableNumber` to a `BsonValue` type. Throws if `value` cannot be 
+    /// exactly represented by an `Int`, `Int32`, `Int64`, or `Double`. 
+    fileprivate func boxNumber<T: CodableNumber>(_ value: T) throws -> BsonValue {
+        guard let number = value.bsonValue else {
+            throw EncodingError._numberError(at: self.codingPath, value: value)
+        }
+        return number
+    }
+
+    fileprivate func box<T: Encodable>(_ value: T) throws -> BsonValue? {
+
+        // if it's already a BsonValue, just return it, unless if it is an 
+        // array. technically [Any] is a BsonValue, but we can only use this
+        // short-circuiting if all the elements are actually BsonValues.
+        if let bsonValue = value as? BsonValue, !(bsonValue is [Any]) {
+            return bsonValue
+        } else if let bsonArray = value as? [BsonValue?] {
```

Comment:
Why `[BsonValue?]` and not `[BsonValue]`?

Reply:
this would come up in the case someone was encoding an array of optionals that had a nil in it. for example
```swift
struct MyStruct: Encodable {
    let ints: [Int?]
}

let s = MyStruct(ints: [1, 2, nil])
```